### DISCUSSION TITLE
Why is it okay to have a system that takes a entity and a const component, but not okay to take an entity and non-const component?

### DISCUSSION THREAD

**Discussion by TypeDefinition** [SanderMertens/flecs#916] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/13763067/213245408-f3c4f49d-95b0-4367-b50b-e708e90d5b1e.png)

**Comment by SanderMertens** [comment, marked as answer]
My guess is that `root` is an empty type, and empty types are passed by value (try `root p`).

The reason for this is that empty types are not actually stored by the ECS, and if they are part of a system/query signature a temporary object of the empty type is created to pass to the `each` function.

The reason this works with const but not for non-const, is that while C++ does let you pass a temporary object to a `const T&`, you can't pass a temporary object to a `T&`.

Since an empty type can't actually store values, a cleaner way is sometimes to use `with`:

```cpp
world.system("Move")
  .with<root>()
  .each([](flecs::entity e) { // component passed by `with` doesn't become part of signature
  });
```

Hope that helps!

**Reply by TypeDefinition** [reply]
Thank you. `root` indeed is an empty type.
